### PR TITLE
Use status as part of aggregation for instances

### PR DESCRIPTION
**Pull request by oswynb**
What it says on the box.

We need this functionality to bill accurately.

@fractalcat @barneydesmond @tranma 
Suggestions of better/nicer way to structure these changes + whether the selected billability of statuses are suitable?

Am thinking this might better belong in borel than here.

### CODE REVIEW

**Comment by tranma** [comment]
is this logic the same as simply filtering out the decoded points with unbillable statuses _before_ doing the fold?

**Comment by oswynb** [comment]
It is the same as filtering them out after the fold. Due to how we calculate the timestamp delta between points, if we filter beforehand we will overbill.

**Comment by oswynb** [comment]
Am thinking even more this logic should belong in borel. @tranma  ?

**Comment by tranma** [comment]
borel can supply a predicate on all decoded payloads, and all folds will need to check the predicate. If this is what we want to do often, that's fine.

**Comment by oswynb** [comment]
What do you think of filtering on the FoldResult instead of within the Fold? That way borel could do it's own munging as required.

**Comment by tranma** [comment]
that's possible, I don't know what's cleaner. It depends on the extent of what you want to do.

In any case I'm trying to clean this up a bit to enforce more structure. The interface types will change slightly, but either approach should still work.

**Comment by oswynb** [comment]
roger

**Comment by tranma** [comment]
If this is to be done from Borel:
- in borel-core, add your predicates to the Env that will be passed to ceilometer-common
- in the `Known` instances for the resources you want to modify (see Ceilometer.Fold), pick the right predicates and give it to the specific fold

File: lib/Ceilometer/Fold.hs [hunk 1]
```diff
@@ -253,15 +253,17 @@ type AGaugePollster x = ( Maybe (Timed x)          -- latest
...
-  = let delta = t2 - t1
-    in  (Just x, M.insertWith (+) (v1 ^. value) (fromIntegral delta) acc)
+          => (x -> Bool) -> AGaugePollster x -> Timed x -> AGaugePollster x
+sGaugePollster _          (Nothing,            acc) x =  (Just x, acc)
+sGaugePollster isBillable (Just (Timed t1 v1), acc) x@(Timed t2 _)
+  = let delta = t2 - t1 in
+    if isBillable v1 then
+      (Just x, M.insertWith (+) (v1 ^. value) (fromIntegral delta) acc)
+    else
+      (Just x, acc)
```
Comment (tranma) [inline lib/Ceilometer/Fold.hs:264]:
```
= (Just x, if billable v1 then .... )
```

File: lib/Ceilometer/Types/Instance.hs [hunk 2]
```diff
@@ -60,6 +62,36 @@ $(declarePF    "Instance"
...
+isInstanceStatusBillable InstancePassword         = True
+isInstanceStatusBillable InstanceResize           = True
+isInstanceStatusBillable InstanceVerifyResize     = True
+isInstanceStatusBillable InstanceRevertResize     = True
+isInstanceStatusBillable InstancePaused           = False
+isInstanceStatusBillable InstanceSuspended        = False
+isInstanceStatusBillable InstanceRescue           = False
+isInstanceStatusBillable InstanceMigrating        = False
+
 data PDInstanceVCPU = PDInstanceVCPU PFInstanceStatus PFValue32
```
Comment (tranma) [inline lib/Ceilometer/Types/Instance.hs:95]: can we not have Java names? billableFlavor, billableStatus, etc. is fine

also, I thought we're moving this mapping to borel?

**Comment by oswynb** [comment]
Changes will be ported to borel/borel-core after revamp

**Pull request closed without merging**